### PR TITLE
[FLINK-1589] Add option to pass configuration to LocalExecutor

### PR DESCRIPTION
**Pull request by rmetzger**
Please review the changes.

I'll add a testcase and update the documentation later today.

### CODE REVIEW

**Comment by rmetzger** [comment]
I've added documentation and tests to the change.
Lets see if travis gives us a green light.

File: flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/ExecutionEnvironmentITCase.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+
+	@Test
+	public void testLocalEnvironmentWithConfig() throws Exception {
+		IllegalArgumentException e = null;
+		try {
+			Configuration conf = new Configuration();
+			conf.setBoolean(ConfigConstants.FILESYSTEM_DEFAULT_OVERWRITE_KEY, true);
+			conf.setString(ConfigConstants.TASK_MANAGER_TMP_DIR_KEY, "/tmp/thelikelyhoodthatthisdirectoryexisitsisreallylow");
```
Comment (tillrohrmann) [inline flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/ExecutionEnvironmentITCase.java]: Are you sure?

**Comment by tillrohrmann** [comment]
I think we should rework the test case to check that the configuration is properly passed to the system. Right now the exception is thrown in ```ds.writeAsText(null)``` because we pass ```null```. 

I'd propose something like @StephanEwen did in the PR #410. We set the number of slots in the configuration and the job to ```PARALLELISM_AUTO_MAX```. With the special input format which produces only a single element per split, we can count the number of parallel tasks, given that every task receives only one input split.

File: flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/ExecutionEnvironmentITCase.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+			conf.setInteger(ConfigConstants.AKKA_ASK_TIMEOUT, 3);
+			final ExecutionEnvironment env = ExecutionEnvironment.createLocalEnvironment(conf);
+
+			DataSet<Integer> ds = CollectionDataSets.getIntegerDataSet(env);
+			ds.writeAsText(null);
+			env.execute();
+		} catch(IllegalArgumentException iae) {
```
Comment (StephanEwen) [inline flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/ExecutionEnvironmentITCase.java]: Why does this cause an `IllegalArgumentException? This seems like the wrong exception type ;-)

File: flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/ExecutionEnvironmentITCase.java (hunk 1 above)
Comment (StephanEwen) [inline flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/ExecutionEnvironmentITCase.java]: To be on the safe side, we used in other tests a temp directory (specially created) where we removed the write permissions for the user (and the in finally re-grant them in order to cleanly remove the directory)

File: flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/ExecutionEnvironmentITCase.java (hunk 2 above)
Comment (rmetzger) [inline flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/ExecutionEnvironmentITCase.java]: Yeah .. Some stuff is messed up in the test. 
So right now the `ds.writeAsText(null);` is causing he `IllegalArgumentException`, but my intention was to get the `IllegalArgumentException` from the lower layers of the systems, namely the Akka utils complaining about the incorrect ask timeout.

I'll follow @tillrohrmann's suggestion to implement the test differently.

**Comment by rmetzger** [comment]
@tillrohrmann: Thank you for the good feedback!
I'll continue working on this once https://github.com/apache/flink/pull/410 is merged to master.

**Comment by rmetzger** [comment]
I've updated the PR. It is now ready for review again.

**Comment by rmetzger** [comment]
If there are no objections, I'm going to merge the PR in the next 24 hours.

File: flink-clients/src/main/java/org/apache/flink/client/LocalExecutor.java [hunk 3]
```diff
@@ -103,6 +110,9 @@ public void start() throws Exception {
 				
 				// create the embedded runtime
 				Configuration configuration = getConfiguration(this);
+				if(this.configuration != null) {
+					configuration.addAll(this.configuration);
+				}
```
Comment (mxm) [inline flink-clients/src/main/java/org/apache/flink/client/LocalExecutor.java]: Wouldn't it be better to move the null check to the constructor and create the default `Configuration` there via a call to `getConfiguration`?

File: flink-clients/src/main/java/org/apache/flink/client/LocalExecutor.java [hunk 4]
```diff
@@ -70,6 +72,11 @@ public LocalExecutor() {
 		}
 	}
 
+	public LocalExecutor(Configuration conf) {
+		super();
+		this.configuration = conf;
+	}
```
Comment (mxm) [inline flink-clients/src/main/java/org/apache/flink/client/LocalExecutor.java]: Should `super()` not be actually `this()`?

File: flink-java/src/main/java/org/apache/flink/api/java/LocalEnvironment.java [hunk 5]
```diff
@@ -32,6 +33,7 @@
  * machine.
  */
 public class LocalEnvironment extends ExecutionEnvironment {
+	private Configuration configuration = null;
```
Comment (StephanEwen) [inline flink-java/src/main/java/org/apache/flink/api/java/LocalEnvironment.java]: The `= null` initializations are unnecessary.

**Comment by StephanEwen** [comment]
Looks good.

It is possible to add a more Unit-Test style test, rather than an Integration test case (starting a full mini cluster) ?

File: flink-clients/src/main/java/org/apache/flink/client/LocalExecutor.java (hunk 3 above)
Comment (uce) [inline flink-clients/src/main/java/org/apache/flink/client/LocalExecutor.java]: +1, I would rename getConfiguration to createConfiguration as well

File: flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java [hunk 6]
```diff
@@ -1058,6 +1059,20 @@ public static LocalEnvironment createLocalEnvironment(int parallelism) {
 		lee.setParallelism(parallelism);
 		return lee;
 	}
+
+	/**
+	 * Creates a {@link LocalEnvironment}. The local execution environment will run the program in a
+	 * multi-threaded fashion in the same JVM as the environment was created in. It will use the
+	 * degree of parallelism specified in the parameter.
```
Comment (uce) [inline flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java]: degree of parllelism has been renamed recently

**Comment by uce** [comment]
Looks good.

- Can you add missing `@Override` tags to `executePlan` and `getOptimizerPlanAsJSON` while editing the file?
- If you only want to ensure that the respective Configuration values are set, we could indeed just get the configuration object and check that the respective values keys are set as expected.

File: flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java (hunk 6 above)
Comment (mxm) [inline flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java]: +1 please change it to parallelism :)

**Comment by rmetzger** [comment]
> It is possible to add a more Unit-Test style test, rather than an Integration test case (starting a full mini cluster) ?

I've chosen this test strategy to be 100% sure that values passed by the user really end up in the system in the end. 
I just checked the time between the first and the last log message, its 643 milliseconds. I think that is okay.

**Comment by rmetzger** [comment]
I'm merging this PR ....

**Pull request closed without merging**